### ISSUE TITLE
ty doesn't recognize deeply inherited base classes/@jsii.implements() decorators

### ISSUE THREAD

**Issue by soderluk** [astral-sh/ty#3200]
### Summary

We rely heavily on AWS CDK, and ty seems to be unable to determine the correct types for a lot of the concrete classes that implements/inherits from protocol classes.

An example:

```python
from aws_cdk import aws_iam as iam

service_principal: iam.IPrincipal = iam.ServicePrincipal("lambda.amazonaws.com)

role: iam.IRole = iam.Role(
    self, 
    "Role", 
    assumed_by=service_principal,
    role_name="MyRole",
)
```
This causes ty to report this as `Object of type 'Role' is not assignable to 'IRole'`. Same with the service principal: `Object of type 'ServicePrincipal' is not assignable to 'IPrincipal'`.

The definition of `Role` is:
```python
@jsii.implements(IRole)
class Role(
    _Resource_45bc6135,
    metaclass=jsii.JSIIMeta,
    jsii_type="aws-cdk-lib.aws_iam.Role",
):
...
```

Another case is the `ServicePrincipal()` which actually implements `IPrincipal` via mixins from the base class, but ty doesn't recognize it.

```python
class ServicePrincipal(
    PrincipalBase,
    metaclass=jsii.JSIIMeta,
    jsii_type="aws-cdk-lib.aws_iam.ServicePrincipal",
):
...
```
```python
@jsii.implements(IAssumeRolePrincipal, IComparablePrincipal)
class PrincipalBase(
    metaclass=jsii.JSIIAbstractClass,
    jsii_type="aws-cdk-lib.aws_iam.PrincipalBase",
):
...
```
```python
@jsii.interface(jsii_type="aws-cdk-lib.aws_iam.IComparablePrincipal")
class IComparablePrincipal(IPrincipal, typing_extensions.Protocol):
...

@jsii.interface(jsii_type="aws-cdk-lib.aws_iam.IAssumeRolePrincipal")
class IAssumeRolePrincipal(IPrincipal, typing_extensions.Protocol):
...
```

It would be great if ty would recognize the `@jsii.implements(IRole)` decorator and report `Role()` as an implementation of `IRole`.

### Version

ty 0.0.26

**Comment by sharkdp** [comment]
> This causes ty to report this as `Object of type 'Role' is not assignable to 'IRole'`.

The `IRole` protocol has this member:
```py
    @jsii.member(jsii_name="grantAssumeRole")
    def grant_assume_role(self, grantee: "IPrincipal") -> "Grant":
```

whereas the `Role` class has this member:
```py
    @jsii.member(jsii_name="grantAssumeRole")
    def grant_assume_role(self, identity: "IPrincipal") -> "Grant":
```

Note that the name of the parameter is not the same, and this parameter is not positional-only, i.e. `grant_assume_role` can be called with `my_irole_instance.grant_assume_role(grantee=…)` which would fail on instances of `Role`.

So this is actually a true positive. We plan to improve our diagnostic in ty. In fact, on my branch [here](https://github.com/astral-sh/ruff/pull/24309), ty emits:
```
error[invalid-assignment]: Object of type `Role` is not assignable to `IRole`
  --> main.py:5:7
   |
 3 |   service_principal: iam.IPrincipal = iam.ServicePrincipal("lambda.amazonaws.com")
 4 |
 5 |   role: iam.IRole = iam.Role(
   |  _______---------___^
   | |       |
   | |       Declared type
 6 | |     None,
 7 | |     "Role",
 8 | |     assumed_by=service_principal,
 9 | |     role_name="MyRole",
10 | | )
   | |_^ Incompatible value of type `Role`
   |
info: type `Role` is not compatible with protocol `IRole`
      ├── protocol member `grant_assume_role` is incompatible
      └── parameter `identity` does not match `grantee` (and can be used as a keyword parameter)
```


So I think this should actually be fixed upstream in `aws_cdk`.

> It would be great if ty would recognize the `@jsii.implements(IRole)` decorator and report `Role()` as an implementation of `IRole`.

I'm not really familiar with `jsii`, but it seems to be related to JavaScript interop, so I don't see a good reason why type checkers should recognize (and act on) this third-party decorator.

**Comment by soderluk** [comment]
Good catch on the grant_assume_role! This is a great addition to the error reporting.
I'll open a ticket on aws_cdk to check and work around this issue until then. Closing this issue.
